File: KFCountDownProject/KFCountDownView.m
//
//  KFCountDownView.m
//  TestProject
//
//  Created by zhkf on 2017/10/24.
//  Copyright © 2017年 zhkf. All rights reserved.
//

#import "KFCountDownView.h"

static const CGFloat kKFCountDownTotalTime = 15;
static const CGFloat kKFTimeInterval = 0.05;
static const CGFloat kKFCircleLineWidth = 4;

@implementation KFCountDownView
{
    CGFloat _currentProgress;
    NSInteger _count;
    dispatch_source_t _gcdTimer;
    CAEmitterLayer *_eLayer;
    CAEmitterCell *_eCell;
    CAShapeLayer *_shapeLayer;
}

- (void)awakeFromNib
{
    [super awakeFromNib];
    [self initializeView];
//    self.backgroundColor = [UIColor redColor];
}

- (instancetype)init
{
    return [self initWithFrame:CGRectZero];
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self initializeView];
    }
    return self;
}

- (void)initializeView
{
    self.backgroundColor = [UIColor clearColor];
    
    self.closeButtonBackgroundColor = [UIColor grayColor];
    self.totalTime = kKFCountDownTotalTime;
    self.countDownCircleColor = [UIColor colorWithRed:102/255.0 green:1 blue:1 alpha:1];
    
    CGFloat width = self.frame.size.width - kKFCircleLineWidth;
    UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
    button.backgroundColor = self.closeButtonBackgroundColor;
    [self addSubview:button];
    button.layer.cornerRadius = width/2;
    button.layer.masksToBounds = YES;
    button.frame = CGRectMake(0, 0, width, width);
    button.center = CGPointMake(self.frame.size.width/2, self.frame.size.height/2);
    [button setBackgroundImage:self.closeButtonImage forState:UIControlStateNormal];
    
    [button addTarget:self action:@selector(closeButtonClick) forControlEvents:UIControlEventTouchUpInside];
    
    _eLayer = [CAEmitterLayer layer];
    _eLayer.frame = CGRectMake(0, 0, width, width);
    
    _eLayer.emitterSize = CGSizeMake(kKFCircleLineWidth, kKFCircleLineWidth);
    _eLayer.emitterShape = kCAEmitterLayerCircle;
    _eLayer.emitterMode = kCAEmitterLayerSurface;
    _eLayer.renderMode = kCAEmitterLayerAdditive;
    
    CAEmitterCell *cell = [CAEmitterCell emitterCell];
    cell.name = @"countdown";
    
    cell.color = _countDownCircleColor.CGColor;
    cell.contents = (__bridge id _Nullable)([self imageWithColor:_countDownCircleColor andSize:CGSizeMake(1, 1)].CGImage);
    cell.birthRate = 0;
    cell.lifetime = 0.6;

    _eLayer.emitterCells = @[cell];
    _eCell = cell;
    
    [self.layer addSublayer:_eLayer];
    
    CAShapeLayer *shapeLayer = [[CAShapeLayer alloc] init];
    UIBezierPath *path = [[UIBezierPath alloc] init];
    [path addArcWithCenter:CGPointMake(self.bounds.size.width/2, self.bounds.size.height/2) radius:self.bounds.size.width/2 startAngle:-M_PI_2 endAngle:M_PI_2 * 2 clockwise:YES];
    shapeLayer.path = path.CGPath;
    shapeLayer.lineJoin = kCALineJoinRound;
    shapeLayer.lineWidth = kKFCircleLineWidth;
    shapeLayer.strokeColor = _countDownCircleColor.CGColor;
    shapeLayer.fillColor = [UIColor clearColor].CGColor;
    [self.layer addSublayer:shapeLayer];
    
    _shapeLayer = shapeLayer;
    
    [self startAnimation];
}

- (void)setCountDownEmitterColor:(UIColor *)countDownEmitterColor
{
    _countDownEmitterColor = countDownEmitterColor;
    [self reloadEmitterColor:countDownEmitterColor];
}

- (void)reloadEmitterColor:(UIColor *)color
{
    _eCell.color = color.CGColor;
    _eCell.contents = (__bridge id _Nullable)([self imageWithColor:color andSize:CGSizeMake(1, 1)].CGImage);
}

- (void)setCountDownCircleColor:(UIColor *)countDownCircleColor
{
    _countDownCircleColor = countDownCircleColor;
    if (!_countDownEmitterColor) {
        self.countDownEmitterColor = countDownCircleColor;
    }
}

-(UIImage*)imageWithColor:(UIColor*)color andSize:(CGSize)size
{
    UIGraphicsBeginImageContext(size);
    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextSetFillColorWithColor(context, color.CGColor);
    
    CGRect rect = CGRectMake(0, 0, size.width, size.height);
    UIBezierPath *bezierPath=[UIBezierPath bezierPathWithRoundedRect:rect cornerRadius:size.width/2.0];
    
    CGContextAddPath(context, bezierPath.CGPath);
    CGContextFillPath(context);
    CGContextSetFillColorWithColor(context, color.CGColor);
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return image;
}

- (void)startAnimation
{
    self.hidden = NO;
    _count = 0;
    _currentProgress = 0;
    
    if (_gcdTimer == nil) {
        _gcdTimer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, dispatch_get_main_queue());
        dispatch_time_t start = dispatch_time(DISPATCH_TIME_NOW, kKFTimeInterval * NSEC_PER_SEC);
        uint64_t interval = (uint64_t)(kKFTimeInterval * NSEC_PER_SEC);
        dispatch_source_set_timer(_gcdTimer, start, interval, kKFTimeInterval/2 * NSEC_PER_SEC);
        
        __weak typeof(self) weakSelf = self;
        dispatch_source_set_event_handler(_gcdTimer, ^{
            [weakSelf timeRun];
        });
    }
    dispatch_resume(_gcdTimer);
}

- (void)endAnimation
{
    dispatch_suspend(_gcdTimer);
    self.hidden = YES;
}

- (void)timeRun
{
    _count = _count + 1;
    _currentProgress = _count / (_totalTime * 1/kKFTimeInterval);
    if (_currentProgress > 1) {
        [self closeButtonClick];
        return;
    }
    if (_eCell.birthRate == 0) {
        _eCell.birthRate = 80;
    }
    
    UIBezierPath *path = [UIBezierPath bezierPath];
    CGFloat passAngle = _currentProgress * M_PI_2 * 4;
    CGPoint arcCenter = CGPointMake(self.center.x - self.frame.origin.x, self.center.y - self.frame.origin.y);
    CGFloat radius = (self.frame.size.width - kKFCircleLineWidth)/2;
    
    [path addArcWithCenter:arcCenter radius:radius startAngle:-M_PI_2 + passAngle endAngle:M_PI_2 * 3 clockwise:YES];
    _shapeLayer.path = path.CGPath;
    
    CGFloat currentPointX = arcCenter.x + cosf(-M_PI_2 + passAngle)*radius;
    CGFloat currentPointY = arcCenter.y + sinf(-M_PI_2 + passAngle)*radius;
    _eLayer.emitterPosition = CGPointMake(currentPointX, currentPointY);
}

- (void)closeButtonClick
{
    [self endAnimation];
}

- (void)dealloc
{
    dispatch_cancel(_gcdTimer);
    _gcdTimer = nil;
}

@end

File: KFCountDownProject/KFCountDownView.swift
//
//  KFCountDownView.swift
//  KFCountDownProject
//
//  Created by zhkf on 2017/12/22.
//  Copyright © 2017年 KF. All rights reserved.
//

import UIKit

class KFCountDownView: UIView {
    
    public var totalTime: Int = 15
    
    /// interval should be equal or greater than 0.001
    public var animationInterval: Float = 0.05
    public var lineWidth: CGFloat = 4
    public var closeButtonImage: UIImage? = nil {
        didSet {
            closedButton?.setBackgroundImage(closeButtonImage, for: .normal)
        }
    }
    public var closeButtonBackgroundColor: UIColor = UIColor.gray {
        didSet {
            closedButton?.backgroundColor = closeButtonBackgroundColor
        }
    }
    
    public var countDownCircleColor = UIColor.init(red: 102/255.0, green: 1, blue: 1, alpha: 1) {
        didSet {
            countDownEmitterColor = countDownCircleColor
        }
    }
    public var countDownEmitterColor: UIColor {
        didSet {
            eCell?.contents = getImageWithColor(color: countDownEmitterColor, size: CGSize (width: 1, height: 1))?.cgImage
        }
    }
    
    private var closedButton: UIButton?
    private var eLayer: CAEmitterLayer?
    private var eCell: CAEmitterCell?
    private var shaperLayer: CAShapeLayer?
    private var gcdTimer: DispatchSourceTimer?
    private var count = 0
    private var currentProgress: Float = 0
    
    override init(frame: CGRect) {
        self.countDownEmitterColor = self.countDownCircleColor
        super.init(frame: frame)
        initializeView()
    }
    
    convenience init() {
        fatalError("not supported")
    }
    
    private func initializeView() {
        self.backgroundColor = UIColor.clear
        
        let width = self.frame.width - lineWidth
        let closedButton = UIButton.init(type: .custom)
        closedButton.backgroundColor  = closeButtonBackgroundColor;
        closedButton.layer.cornerRadius = width / 2
        closedButton.layer.masksToBounds = true
        closedButton.frame = CGRect (x: 0, y: 0, width: width, height: width)
        closedButton.center = CGPoint (x: self.frame.width/2, y: self.frame.height/2)
        self.addSubview(closedButton)
        
        closedButton .addTarget(self, action: #selector(closedButtonClick), for: .touchUpInside)
        self.closedButton = closedButton
        
        let layer = CAEmitterLayer.init()
        layer.frame = CGRect (x: 0, y: 0, width: width, height: width)
        layer.emitterSize = CGSize (width: lineWidth, height: lineWidth)
        layer.emitterShape = kCAEmitterLayerCircle
        layer.emitterMode = kCAEmitterLayerSurface
        layer.renderMode = kCAEmitterLayerAdditive
        
        eLayer = layer
        
        let cell = CAEmitterCell.init()
        cell.name = "countDown"
        cell.color = countDownEmitterColor.cgColor
        cell.contents = getImageWithColor(color: countDownEmitterColor, size: CGSize (width: 1, height: 1))?.cgImage
        cell.birthRate = 200
        cell.lifetime = 0.3
        
        eLayer?.emitterCells = [cell]
        eCell = cell
        
        self.layer .addSublayer(layer)
        
        let shapeLayer = CAShapeLayer.init()
        
        shapeLayer.path = getCirclePath(progress: 0).0.cgPath
        shapeLayer.lineJoin = kCALineJoinRound
        shapeLayer.lineWidth = lineWidth
        shapeLayer.strokeColor = countDownCircleColor.cgColor
        shapeLayer.fillColor = UIColor.clear.cgColor
        self.layer .addSublayer(shapeLayer)
        
        self.shaperLayer = shapeLayer
        
        startAnimation()
    }
    
    @objc func closedButtonClick() {
        endAnimation()
    }
    
    private func startAnimation() {
        self.isHidden = false
        count = 0
        currentProgress = 0
        if gcdTimer == nil {
           createTimer()
        }
        gcdTimer?.resume()
    }
    
    private func createTimer() {
        gcdTimer = DispatchSource.makeTimerSource(queue: DispatchQueue.main)
        let timeInterval = DispatchTimeInterval.milliseconds(Int(animationInterval * 1000))
        let leewayInterval = DispatchTimeInterval.milliseconds(Int(animationInterval * 500))
        
        gcdTimer?.schedule(deadline: .now() + timeInterval, repeating: timeInterval, leeway: leewayInterval)
        gcdTimer?.setEventHandler(handler: { [weak self] in
            self? .timeRun()
        })
    }
    
    private func timeRun() {
        count = count + 1
        currentProgress = Float(count) * animationInterval/Float(totalTime)
        if currentProgress > 1 {
            closedButtonClick()
            return
        }
        let tuple = getCirclePath(progress: currentProgress)
        shaperLayer?.path = tuple.0.cgPath
        eLayer?.emitterPosition = tuple.1
    }
    
    private func endAnimation() {
        gcdTimer?.suspend()
        count = 0
        currentProgress = 0
        
        self.isHidden = true
    }
    
    private func getCirclePath(progress: Float) -> (UIBezierPath, CGPoint) {
        let path = UIBezierPath.init()
        let passAngle = progress * Float.pi * 2
        let arcCenter = CGPoint (x: self.center.x - self.frame.minX, y: self.center.y - self.frame.minY)
        let radius = self.frame.width/2 - lineWidth/2
        path.addArc(withCenter: arcCenter, radius: radius, startAngle: CGFloat(-Float.pi/2 + passAngle), endAngle: CGFloat(Float.pi * 1.5), clockwise: true)
        
        let currentPointX: CGFloat = arcCenter.x + CGFloat(cosf(-Float.pi/2 + passAngle)) * radius
        let currentPointY: CGFloat = arcCenter.y + CGFloat(sinf(-Float.pi/2 + passAngle)) * radius
        return (path, CGPoint (x: currentPointX, y: currentPointY))
    }
    
    private func getImageWithColor(color: UIColor, size: CGSize) -> UIImage? {
        UIGraphicsBeginImageContext(size)
        let context = UIGraphicsGetCurrentContext()
        
        let rect = CGRect (x: 0, y: 0, width: size.width, height: size.height)
        let bezierPath = UIBezierPath.init(roundedRect: rect, byRoundingCorners: .allCorners, cornerRadii: CGSize (width: size.width/2, height: size.height/2))
        context?.addPath(bezierPath.cgPath)
        context?.setFillColor(color.cgColor)
        context?.setStrokeColor(color.cgColor)
        context?.fillPath()
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        return image
    }
    
    required init?(coder aDecoder: NSCoder) {
        self.countDownEmitterColor = self.countDownCircleColor
        super.init(coder: aDecoder)
        initializeView()
    }
    
    deinit {
        gcdTimer?.cancel()
        gcdTimer = nil
    }
}
